[git-webkit] Accidental overwrite of pull request with similar branch names in Github

JonWBedard opened the pull request:
#### 9eedf00c442768bab889b5e7ca6bc3e24c07bb53
<pre>
[git-webkit] Accidental overwrite of pull request with similar branch names in Github
<a href="https://bugs.webkit.org/show_bug.cgi?id=244660">https://bugs.webkit.org/show_bug.cgi?id=244660</a>
&lt;rdar://problem/99648208&gt;

Reviewed by Elliott Williams and Ryan Haddad.

* Tools/Scripts/libraries/webkitscmpy/setup.py: Bump version.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/__init__.py: Ditto.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/contributor.py:
(Contributor.__repr__): Print ? if &apos;email&apos; is &apos;None&apos;.
(Contributor.__cmp__): Consider other string representations of a Contributor when comparing against a string.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/local/git.py:
(Git.remote): Pass contributors object so that Contributors have their actual names.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/pull_request.py:
(PullRequest.find_existing_pull_request): Prefer pull-requests owned by the current user if branch names collide.
(PullRequest.create_pull_request): If the user provides a branch name already used by another user, offer to
push to that user&apos;s branch instead of the current user&apos;s branch. If the user opts out of this, create a new pull-request.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/remote/bitbucket.py:
(BitBucket.credentials): Accept options GitHub&apos;s credential&apos;s function accepts.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/test/land_unittest.py:
(TestLandGitHub): Specify current user.

Canonical link: <a href="https://commits.webkit.org/254478@main">https://commits.webkit.org/254478@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/0d7d0d2622c816a5fa3603bc43ae6e9b75e3db48

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/89196 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/69/builds/33757 "Built successfully") | [  ~~🛠 mac~~](https://ews-build.webkit.org/#/builders/43/builds/20016 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/8/builds/98515 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/154828 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/93205 "Passed tests") | [  ~~🛠 ios-sim~~](https://ews-build.webkit.org/#/builders/64/builds/32265 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac-debug~~](https://ews-build.webkit.org/#/builders/71/builds/27813 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/36/builds/81558 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/92984 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/94843 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/64/builds/32265 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/61/builds/76121 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/81558 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/5/builds/92783 "Passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/64/builds/32265 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/43/builds/20016 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/36/builds/81558 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/66/builds/30044 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/43/builds/20016 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🧪 services](https://ews-build.webkit.org/#/builders/20/builds/88252 "Passed tests") | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/67/builds/29768 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/43/builds/20016 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/63/builds/33215 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/61/builds/76121 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/1319 "Built successfully and passed tests") | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/65/builds/31916 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/43/builds/20016 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
<!--EWS-Status-Bubble-End-->

emw-apple commented on Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/pull_request.py:
```diff
@@ -354,16 +357,47 @@ def create_pull_request(cls, repository, args, branch_point, callback=None, unbl
+                if not pr_target.startswith('{}/'.format(user)):
+                    target = pr_target.split('/')[0]
+                    if '-' in pr_target:
+                        target = '{}-{}'.format(target, pr_target.split('-')[-1])
```
What is this doing? If pr_target is `emw-apple/WebKit`, it seems like `target` would be `emw-apple-apple/WebKit`.

JonWBedard replied:
This is trying to handle the case of `emw-apple/WebKit-security`, neglected to consider the case where someone's username contains `-`, will fix.

emw-apple commented on Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/pull_request.py:
```diff
@@ -354,16 +357,47 @@ def create_pull_request(cls, repository, args, branch_point, callback=None, unbl
+            if existing_pr and user and existing_pr.author != user and (args.defaults or Terminal.choose(
+                "'{}' is owned by '{}'\nWould you like to take over this pull-request?".format(existing_pr, existing_pr.author),
+                default='No',
+            ) == 'No'):
```
FWIW, I tried this locally and saw what appear to be some line control glitches: 
<img width="1223" alt="Screen Shot 2022-09-07 at 16 57 33" src="https://user-images.githubusercontent.com/96081375/189005425-5fdd56f0-f840-452d-b2b0-a4b150b965f7.png">
Maybe because the prompt message doesn't end in a newline?

JonWBedard replied:
Curious if you see this in a new terminal session?

I think I need to fix this in `Terminal.choose`, because if it's a problem here, there are a couple of other places it might be a problem as well.

JonWBedard replied:
Attempting a fix for this in #4131

emw-apple reviewed (changes requested):
Is there a reason we want to force unique branch names for all our PRs for all time? That feels like an unnecessary limitation to me. As I see it, part of the advantage of fork-based workflows is that all contributors don't share the same namespace for branches.

emw-apple commented on Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/pull_request.py:
```diff
@@ -354,16 +357,47 @@ def create_pull_request(cls, repository, args, branch_point, callback=None, unbl
+            user, _ = remote_repo.credentials(required=False)
+
             log.info("Checking if PR already exists...")
             existing_pr = cls.find_existing_pull_request(repository, remote_repo)
```
If I take over a PR (by pushing to the author's branch), will this call notice my coauthorship, or will I have to go through the "Would you like to take over this PR?" prompt again?

JonWBedard replied:
It should assign the PR to you, so you won't get the prompt.

When working on this PR, I discovered that most contributors probably haven't given others write access to their branches, so if you fail to push, we won't update the PR and then you'll get the prompt again, but that seems appropriate.

emw-apple commented on Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/pull_request.py:
```diff
@@ -354,16 +355,41 @@ def create_pull_request(cls, repository, args, branch_point, callback=None, unbl
+            if existing_pr and user and existing_pr.author != user and (args.defaults or Terminal.choose(
+                "'{}' is owned by '{}'\nWould you like to take over this pull-request?".format(existing_pr, existing_pr.author),
+                default='No',
+            ) == 'No'):
```
UI nit: Could we better explain that if you _don't_ take over, it'll create a new PR for you? Something like:

    Take over this pull-request instead of creating a new one? (Yes/[No]):

I don't want people to think their only option is to take over the PR or cancel.

We could even switch from Y/N prompts to action verbs. Something like:

    You can either:
      [C]reate a new pull-request
      [O]verwrite {PR} and assign yourself
    How should git-webkit proceed? ([C]reate/[o]verwrite)

emw-apple commented on Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/pull_request.py:
```diff
@@ -354,16 +355,41 @@ def create_pull_request(cls, repository, args, branch_point, callback=None, unbl
+                default='No',
+            ) == 'No'):
+                if target == source_remote:
+                    sys.stderr.write("'{}' already exists on '{}', creating a pull-request would overwrite it\n".format(
+                        repository.branch, target,
+                    ))
+                    return 1
```
How is this reachable? Can `source_remote` ever be a fork URL?

JonWBedard replied:
This is reachable if you aren't using a `fork` workflow, which is true in certain BitBucket configurations `git-webkit` also supports.

JonWBedard committed:
[git-webkit] Accidental overwrite of pull request with similar branch names in Github
https://bugs.webkit.org/show_bug.cgi?id=244660
<rdar://problem/99648208>

Reviewed by Elliott Williams and Ryan Haddad.

* Tools/Scripts/libraries/webkitscmpy/setup.py: Bump version.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/__init__.py: Ditto.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/contributor.py:
(Contributor.__repr__): Print ? if 'email' is 'None'.
(Contributor.__cmp__): Consider other string representations of a Contributor when comparing against a string.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/local/git.py:
(Git.remote): Pass contributors object so that Contributors have their actual names.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/pull_request.py:
(PullRequest.find_existing_pull_request): Prefer pull-requests owned by the current user if branch names collide.
(PullRequest.create_pull_request): If the user provides a branch name already used by another user, offer to
push to that user's branch instead of the current user's branch. If the user opts out of this, create a new pull-request.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/remote/bitbucket.py:
(BitBucket.credentials): Accept options GitHub's credential's function accepts.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/test/land_unittest.py:
(TestLandGitHub): Specify current user.

Canonical link: https://commits.webkit.org/254478@main
